app/src/test/java/com/training/trainingapp/view_model/SignUpViewModelTest.kt

```diff
@@ -0,0 +1,107 @@
+import kotlinx.coroutines.flow.toList
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.test.UnconfinedTestDispatcher
+import kotlinx.coroutines.test.advanceUntilIdle
+import kotlinx.coroutines.test.runTest
+import kotlinx.coroutines.test.setMain
+import org.junit.After
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertFalse
+import org.junit.Assert.assertTrue
+import org.junit.Before
+import org.junit.Test
+
+@OptIn(ExperimentalCoroutinesApi::class, DelicateCoroutinesApi::class)
+class SignUpViewModelTest {
+    lateinit var sampleViewModel: SignUpViewModel
+
+    val signUpViewUseCase = mockk<SignUpViewUseCase>()
+
+    @Before
+    fun initValues() {
+        Dispatchers.setMain(UnconfinedTestDispatcher())
+        sampleViewModel = SignUpViewModel(signUpViewUseCase)
+    }
+
+    @Test
+    fun GivenEmail_WhenEmailChanged_ThenViewModelEmailMustBeChanged() {
+        var email = "paria.m7616@gmial.com"
+        sampleViewModel.onEmailChanged(email)
+        assertEquals(sampleViewModel.pageState.value.email, email)
+
+        email = "paria"
+        sampleViewModel.onEmailChanged(email)
+        assertEquals(sampleViewModel.pageState.value.email, email)
+
+        email = ""
+        sampleViewModel.onEmailChanged(email)
+        assertEquals(sampleViewModel.pageState.value.email, email)
+    }
+
+    @Test
+    fun GivenCorrectEmail_WhenEmailChangedAndSubmitClicked_ThenViewModelEmailValidateShouldBeTrue() {
+        coEvery { signUpViewUseCase.invoke(any()) } returns SignUpResponse(true)
+
+        val email = "paria.m7616@gmial.com"
+        sampleViewModel.onEmailChanged(email)
+        sampleViewModel.onSubmitButtonClicked()
+        assertTrue(sampleViewModel.pageState.value.emailValidate)
+    }
+
+    @Test
+    fun GivenWrongEmail_WhenEmailChangedAndSubmitClicked_ThenViewModelEmailValidateShouldBeFalse() {
+        val email = "paria."
+        sampleViewModel.onEmailChanged(email)
+        sampleViewModel.onSubmitButtonClicked()
+        assertFalse(sampleViewModel.pageState.value.emailValidate)
+    }
+
+    @Test
+    fun GivenEmptyEmail_WhenEmailChangedAndSubmitClicked_ThenValidationShouldBeFalse() {
+        val email = ""
+        sampleViewModel.onEmailChanged(email)
+        sampleViewModel.onSubmitButtonClicked()
+        assertFalse(sampleViewModel.pageState.value.emailValidate)
+    }
+
+    @Test
+    fun GivenCorrectEmail_WhenEmailChangedAndSubmitClicked_ThenCheckPageState() = runTest {
+        coEvery { signUpViewUseCase.invoke(any()) } returns SignUpResponse(true)
+        val email = "paria.m7616@gmial.com"
+
+        val testResults = mutableListOf<SignUpPageState>()
+        backgroundScope.launch(UnconfinedTestDispatcher(testScheduler)) {
+            sampleViewModel.pageState.toList(testResults)
+        }
+
+        sampleViewModel.onEmailChanged(email)
+        sampleViewModel.onSubmitButtonClicked()
+        advanceUntilIdle()
+
+        assertEquals(4, testResults.size)
```

Comment:
I did not understand what you mean by 4 here

Reply:
In this function we want to get all state changes and check it for correct changing states.
at first our state must be IDLE , after  "sampleViewModel.onEmailChanged(email)" our state are IDLE but email must be set to email value, after "sampleViewModel.onSubmitButtonClicked()" our state change to LOADING and after geting response from usecase out state must be SUCCESS or FAILED and it depends on out mock object

Reply:
We want wait until all tasks be completed, because of that we use advanceUntilIdle.